File: implementations/csharp/client/tests/Specus.Client.Tests/PeerEgressSocketBindingVectorTests.cs
using System.Text.Json;
using Specus.Client.PeerMesh;

namespace Specus.Client.Tests;

/// <summary>
/// Replays <c>peer-egress-socket-binding-v1.json</c> against the .NET choice of interface for an
/// egress socket and the two platforms' table readings.
/// </summary>
/// <remarks>
/// The expectations come from an independent reference in
/// <c>tools/protocol/generate_peer_egress_socket_binding_vectors.py</c>, whose Windows tables are
/// built from the SDK's structure layouts with ctypes and whose macOS tables are sampled captures.
/// </remarks>
public class PeerEgressSocketBindingVectorTests
{
    private static JsonDocument Vector()
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        while (directory is not null)
        {
            var candidate = Path.Combine(directory.FullName, "protocol", "test-vectors",
                "peer-egress-socket-binding-v1.json");
            if (File.Exists(candidate))
            {
                return JsonDocument.Parse(File.ReadAllText(candidate));
            }
            directory = directory.Parent;
        }
        throw new FileNotFoundException("cannot locate peer-egress-socket-binding-v1.json");
    }

    private static List<PeerEgressBindRoute> Routes(JsonElement array) =>
        array.EnumerateArray().Select(route => new PeerEgressBindRoute(
            route.GetProperty("prefix").GetString()!,
            route.GetProperty("interface").GetString()!,
            route.TryGetProperty("gateway", out var gateway) ? gateway.GetString()! : "",
            route.GetProperty("metric").GetInt64(),
            route.GetProperty("usable").GetBoolean())).ToList();

    private static void AssertChoice(string name, IReadOnlyList<PeerEgressBindRoute> routes, JsonElement testCase)
    {
        var expected = testCase.GetProperty("expect").GetProperty("interface");
        var chosen = PeerEgressSocketBinding.Select(routes, testCase.GetProperty("tunnel").GetString(),
            testCase.GetProperty("destination").GetString()!);
        if (expected.ValueKind == JsonValueKind.Null)
        {
            Assert.True(chosen is null, $"{name}: chose {chosen}, want the dial refused");
        }
        else
        {
            Assert.True(expected.GetString() == chosen, $"{name}: chose {chosen}, want {expected.GetString()}");
        }
    }

    [Fact]
    public void SelectionMatchesTheSharedVector()
    {
        using var vector = Vector();
        var cases = vector.RootElement.GetProperty("select").GetProperty("cases");
        Assert.True(cases.GetArrayLength() > 0, "no selection cases");
        foreach (var testCase in cases.EnumerateArray())
        {
            AssertChoice(testCase.GetProperty("name").GetString()!, Routes(testCase.GetProperty("routes")), testCase);
        }
    }

    [Fact]
    public void WindowsTablesMatchTheSharedVector()
    {
        using var vector = Vector();
        var windows = vector.RootElement.GetProperty("windows");
        var forwardByName = new Dictionary<string, List<PeerEgressWindowsForwardRow>>();
        foreach (var table in windows.GetProperty("forwardTables").EnumerateArray())
        {
            var name = table.GetProperty("name").GetString()!;
            var rows = PeerEgressSocketBinding.ParseWindowsForwardTable(
                Convert.FromHexString(table.GetProperty("hex").GetString()!));
            Assert.True(table.GetProperty("expect").GetProperty("parsed").GetBoolean() == rows is not null,
                $"forward {name}: parsed flag");
            if (rows is null)
            {
                continue;
            }
            var expected = table.GetProperty("expect").GetProperty("rows").EnumerateArray()
                .Select(row => new PeerEgressWindowsForwardRow(row.GetProperty("interfaceIndex").GetInt64(),
                    row.GetProperty("prefix").GetString()!, row.GetProperty("nextHop").GetString()!,
                    row.GetProperty("metric").GetInt64()))
                .ToList();
            Assert.Equal(expected, rows);
            forwardByName[name] = rows;
        }
        var interfacesByName = new Dictionary<string, List<PeerEgressWindowsInterfaceRow>>();
        foreach (var table in windows.GetProperty("interfaceTables").EnumerateArray())
        {
            var name = table.GetProperty("name").GetString()!;
            var rows = PeerEgressSocketBinding.ParseWindowsInterfaceTable(
                Convert.FromHexString(table.GetProperty("hex").GetString()!));
            Assert.True(table.GetProperty("expect").GetProperty("parsed").GetBoolean() == rows is not null,
                $"interfaces {name}: parsed flag");
            if (rows is null)
            {
                continue;
            }
            var expected = table.GetProperty("expect").GetProperty("rows").EnumerateArray()
                .Select(row => new PeerEgressWindowsInterfaceRow(row.GetProperty("interfaceIndex").GetInt64(),
                    row.GetProperty("metric").GetInt64(), row.GetProperty("connected").GetBoolean(),
                    row.GetProperty("disableDefaultRoutes").GetBoolean()))
                .ToList();
            Assert.Equal(expected, rows);
            interfacesByName[name] = rows;
        }

        var joined = windows.GetProperty("routes");
        var routes = PeerEgressSocketBinding.WindowsRoutes(
            forwardByName[joined.GetProperty("forward").GetString()!],
            interfacesByName[joined.GetProperty("interfaces").GetString()!]);
        Assert.Equal(Routes(joined.GetProperty("expect")), routes);
        foreach (var testCase in windows.GetProperty("cases").EnumerateArray())
        {
            AssertChoice("windows " + testCase.GetProperty("name").GetString(), routes, testCase);
        }
    }

    [Fact]
    public void MacosTablesMatchTheSharedVector()
    {
        using var vector = Vector();
        var macos = vector.RootElement.GetProperty("macos");
        var byTable = new Dictionary<string, List<PeerEgressBindRoute>>();
        foreach (var entry in macos.GetProperty("routes").EnumerateArray())
        {
            var table = entry.GetProperty("table").GetString()!;
            var routes = PeerEgressSocketBinding.MacosRoutes(macos.GetProperty("tables").GetProperty(table).GetString());
            Assert.Equal(Routes(entry.GetProperty("expect")), routes);
            byTable[table] = routes;
        }
        foreach (var testCase in macos.GetProperty("cases").EnumerateArray())
        {
            AssertChoice("macos " + testCase.GetProperty("name").GetString(),
                byTable[testCase.GetProperty("table").GetString()!], testCase);
        }
    }

    [Fact]
    public void LinuxTablesMatchTheSharedVector()
    {
        using var vector = Vector();
        var linux = vector.RootElement.GetProperty("linux");
        Assert.True(linux.GetProperty("routes").GetArrayLength() > 0, "no Linux tables");
        foreach (var entry in linux.GetProperty("routes").EnumerateArray())
        {
            var table = entry.GetProperty("table").GetString()!;
            Assert.Equal(Routes(entry.GetProperty("expect")),
                PeerEgressRouteCommands.ParseRouteTable(linux.GetProperty("tables").GetProperty(table).GetString()));
        }
        // The captures that justify reading the table at all: `ip route get` answering with the tunnel.
        foreach (var capture in linux.GetProperty("captures").EnumerateObject())
        {
            if (!capture.Name.StartsWith("get-covered", StringComparison.Ordinal))
            {
                continue;
            }
            var hop = PeerEgressRouteCommands.ParseRouteGet(capture.Value.GetProperty("stdout").GetString());
            Assert.True(hop is { } parsed && PeerEgressRouteCommands.HopIsDevice(parsed, "specus0"),
                $"{capture.Name} parsed as {hop}");
        }
    }

    [Fact]
    public void BypassHopsMatchTheSharedVector()
    {
        using var vector = Vector();
        var root = vector.RootElement;
        var tables = new Dictionary<string, Dictionary<string, List<PeerEgressBindRoute>>>
        {
            ["linux"] = root.GetProperty("linux").GetProperty("tables").EnumerateObject()
                .ToDictionary(table => table.Name, table => PeerEgressRouteCommands.ParseRouteTable(table.Value.GetString())),
            ["macos"] = root.GetProperty("macos").GetProperty("tables").EnumerateObject()
                .ToDictionary(table => table.Name, table => PeerEgressSocketBinding.MacosRoutes(table.Value.GetString())),
        };
        var windows = root.GetProperty("windows");
        var forward = windows.GetProperty("forwardTables").EnumerateArray()
            .Where(table => table.GetProperty("name").GetString() == windows.GetProperty("routes").GetProperty("forward").GetString())
            .Select(table => PeerEgressSocketBinding.ParseWindowsForwardTable(Convert.FromHexString(table.GetProperty("hex").GetString()!)))
            .Single();
        var interfaces = windows.GetProperty("interfaceTables").EnumerateArray()
            .Where(table => table.GetProperty("name").GetString() == windows.GetProperty("routes").GetProperty("interfaces").GetString())
            .Select(table => PeerEgressSocketBinding.ParseWindowsInterfaceTable(Convert.FromHexString(table.GetProperty("hex").GetString()!)))
            .Single();
        tables["windows"] = new() { ["typical"] = PeerEgressSocketBinding.WindowsRoutes(forward!, interfaces!) };

        var cases = root.GetProperty("hops").GetProperty("cases");
        Assert.True(cases.GetArrayLength() > 0, "no hop cases");
        foreach (var testCase in cases.EnumerateArray())
        {
            var name = testCase.GetProperty("name").GetString();
            var platform = testCase.GetProperty("platform").GetString()!;
            var routes = platform == "inline"
                ? Routes(testCase.GetProperty("routes"))
                : tables[platform][testCase.GetProperty("table").GetString()!];
            var owned = testCase.GetProperty("owned").EnumerateArray().Select(prefix => prefix.GetString()!).ToList();
            var hop = PeerEgressSocketBinding.SelectBypassHop(routes, testCase.GetProperty("tunnel").GetString(),
                owned, testCase.GetProperty("destination").GetString()!);
            var expected = testCase.GetProperty("expect").GetProperty("hop");
            if (expected.ValueKind == JsonValueKind.Null)
            {
                Assert.True(hop is null, $"{name}: chose {hop}, want no hop");
            }
            else
            {
                Assert.True(hop is { } chosen
                        && chosen.Interface == expected.GetProperty("interface").GetString()
                        && chosen.Gateway == expected.GetProperty("gateway").GetString(),
                    $"{name}: chose {hop}, want {expected}");
            }
        }
    }

    /// <summary>
    /// The fallback a commander takes when its query answers with the tunnel: the table with the
    /// tunnel left out, and a refusal that says why when nothing else leads there.
    /// </summary>
    /// <summary>Every sampled table in the vector, as this runtime parses it.</summary>
    private static Dictionary<string, Dictionary<string, List<PeerEgressBindRoute>>> Tables(JsonElement root)
    {
        var tables = new Dictionary<string, Dictionary<string, List<PeerEgressBindRoute>>>
        {
            ["linux"] = root.GetProperty("linux").GetProperty("tables").EnumerateObject()
                .ToDictionary(table => table.Name, table => PeerEgressRouteCommands.ParseRouteTable(table.Value.GetString())),
            ["macos"] = root.GetProperty("macos").GetProperty("tables").EnumerateObject()
                .ToDictionary(table => table.Name, table => PeerEgressSocketBinding.MacosRoutes(table.Value.GetString())),
        };
        var windows = root.GetProperty("windows");
        var forward = windows.GetProperty("forwardTables").EnumerateArray()
            .Where(table => table.GetProperty("name").GetString() == windows.GetProperty("routes").GetProperty("forward").GetString())
            .Select(table => PeerEgressSocketBinding.ParseWindowsForwardTable(Convert.FromHexString(table.GetProperty("hex").GetString()!)))
            .Single();
        var interfaces = windows.GetProperty("interfaceTables").EnumerateArray()
            .Where(table => table.GetProperty("name").GetString() == windows.GetProperty("routes").GetProperty("interfaces").GetString())
            .Select(table => PeerEgressSocketBinding.ParseWindowsInterfaceTable(Convert.FromHexString(table.GetProperty("hex").GetString()!)))
            .Single();
        tables["windows"] = new() { ["typical"] = PeerEgressSocketBinding.WindowsRoutes(forward!, interfaces!) };
        return tables;
    }

    /// <summary>
    /// What the consumer's own routes look like after a network change, and the repair each needs.
    /// The policy -- a tunnel route on another interface is a conflict, a bypass follows the hop the
    /// table would choose with the owned prefixes left out -- has to be the same in three runtimes.
    /// </summary>
    [Fact]
    public void DriftMatchesTheSharedVector()
    {
        using var vector = Vector();
        var root = vector.RootElement;
        var tables = Tables(root);
        var cases = root.GetProperty("drift").GetProperty("cases");
        Assert.True(cases.GetArrayLength() > 0, "no drift cases");
        foreach (var testCase in cases.EnumerateArray())
        {
            var name = testCase.GetProperty("name").GetString();
            var platform = testCase.GetProperty("platform").GetString()!;
            var routes = platform == "inline"
                ? Routes(testCase.GetProperty("routes"))
                : tables[platform][testCase.GetProperty("table").GetString()!];
            var owned = testCase.GetProperty("owned").EnumerateArray()
                .Select(entry => new PeerEgressRoute(entry.GetProperty("cidr").GetString()!,
                    PeerEgressRoutePlanner.KindFromWireName(entry.GetProperty("kind").GetString()), string.Empty))
                .ToList();
            var found = PeerEgressSocketBinding.Drifts(owned, routes, testCase.GetProperty("tunnel").GetString()!);
            var expect = testCase.GetProperty("expect").EnumerateArray().ToList();
            Assert.True(expect.Count == found.Count, $"{name}: found {string.Join(", ", found)}");
            for (var index = 0; index < expect.Count; index++)
            {
                Assert.True(expect[index].GetProperty("cidr").GetString() == found[index].Route.Cidr
                    && expect[index].GetProperty("reason").GetString() == found[index].Reason
                    && expect[index].GetProperty("action").GetString() == found[index].Action,
                    $"{name}: drift {index} = {found[index]}, want {expect[index]}");
            }
        }
    }

    [Fact]
    public void TheHopFromTheTableLeavesTheTunnelOut()
    {
        using var vector = Vector();
        var routes = PeerEgressRouteCommands.ParseRouteTable(
            vector.RootElement.GetProperty("linux").GetProperty("tables").GetProperty("show-rich").GetString());

        var hop = PeerEgressSocketBinding.BypassHopFromTable("203.0.113.9", "specus0", () => routes);
        Assert.Equal("eth0", hop.Interface);
        Assert.Equal("192.168.64.1", hop.Gateway);
        Assert.Throws<PeerEgressNoPhysicalRouteException>(
            () => PeerEgressSocketBinding.BypassHopFromTable("192.0.2.9", "specus0", () => routes));
        var unreadable = Assert.Throws<IOException>(() => PeerEgressSocketBinding.BypassHopFromTable(
            "203.0.113.9", "specus0", () => throw new InvalidOperationException("ip: not found")));
        Assert.Contains("ip: not found", unreadable.Message);
    }

    [Fact]
    public void SocketOptionsMatchTheSharedVector()
    {
        using var vector = Vector();
        var options = vector.RootElement.GetProperty("socketOptions");
        Assert.Equal(PeerEgressSocketBinding.WindowsIpUnicastIf, options.GetProperty("windows").GetProperty("name").GetInt32());
        Assert.Equal(PeerEgressSocketBinding.MacosIpBoundIf, options.GetProperty("macos").GetProperty("name").GetInt32());
        Assert.Equal(PeerEgressSocketBinding.IpProtoIp, options.GetProperty("windows").GetProperty("level").GetInt32());
        Assert.Equal(PeerEgressSocketBinding.IpProtoIp, options.GetProperty("macos").GetProperty("level").GetInt32());
        foreach (var testCase in options.GetProperty("windows").GetProperty("cases").EnumerateArray())
        {
            Assert.Equal(testCase.GetProperty("hex").GetString(), Convert.ToHexStringLower(
                PeerEgressSocketBinding.WindowsUnicastInterfaceOption(testCase.GetProperty("index").GetInt64())));
        }
        foreach (var testCase in options.GetProperty("macos").GetProperty("cases").EnumerateArray())
        {
            Assert.Equal(testCase.GetProperty("hex").GetString(), Convert.ToHexStringLower(
                PeerEgressSocketBinding.MacosBoundInterfaceOption(testCase.GetProperty("index").GetInt64())));
        }
    }
}

File: implementations/java/client/src/test/java/com/theshuai/specusclient/peer/PeerEgressSocketBindingVectorTests.java
package com.theshuai.specusclient.peer;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HexFormat;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * Replays {@code peer-egress-socket-binding-v1.json} against the Java choice of interface for an
 * egress socket and the two platforms' table readings.
 *
 * <p>The expectations come from an independent reference in
 * {@code tools/protocol/generate_peer_egress_socket_binding_vectors.py}, whose Windows tables are
 * built from the SDK's structure layouts with ctypes and whose macOS tables are sampled captures.
 */
class PeerEgressSocketBindingVectorTests {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static JsonNode vector() throws IOException {
        Path directory = Path.of("").toAbsolutePath();
        for (int depth = 0; depth < 8 && directory != null; depth++) {
            Path candidate = directory.resolve("protocol").resolve("test-vectors")
                    .resolve("peer-egress-socket-binding-v1.json");
            if (Files.exists(candidate)) {
                return MAPPER.readTree(Files.readString(candidate));
            }
            directory = directory.getParent();
        }
        throw new IOException("cannot locate peer-egress-socket-binding-v1.json");
    }

    private static List<PeerEgressSocketBinding.Route> routes(JsonNode array) {
        List<PeerEgressSocketBinding.Route> routes = new ArrayList<>();
        for (JsonNode route : array) {
            routes.add(new PeerEgressSocketBinding.Route(route.path("prefix").asText(),
                    route.path("interface").asText(), route.path("gateway").asText(""),
                    route.path("metric").asLong(), route.path("usable").asBoolean()));
        }
        return routes;
    }

    private static void assertChoice(String name, List<PeerEgressSocketBinding.Route> routes, JsonNode testCase) {
        JsonNode expected = testCase.path("expect").path("interface");
        String chosen = PeerEgressSocketBinding.select(routes, testCase.path("tunnel").asText(),
                testCase.path("destination").asText());
        if (expected.isNull()) {
            assertNull(chosen, name + ": want the dial refused");
        } else {
            assertEquals(expected.asText(), chosen, name);
        }
    }

    @Test
    void selectionMatchesTheSharedVector() throws IOException {
        JsonNode cases = vector().path("select").path("cases");
        assertTrue(cases.size() > 0, "no selection cases");
        for (JsonNode testCase : cases) {
            assertChoice(testCase.path("name").asText(), routes(testCase.path("routes")), testCase);
        }
    }

    @Test
    void windowsTablesMatchTheSharedVector() throws IOException {
        JsonNode windows = vector().path("windows");
        Map<String, List<PeerEgressSocketBinding.WindowsForwardRow>> forwardByName = new HashMap<>();
        for (JsonNode table : windows.path("forwardTables")) {
            String name = table.path("name").asText();
            var rows = PeerEgressSocketBinding.parseWindowsForwardTable(
                    HexFormat.of().parseHex(table.path("hex").asText()));
            assertEquals(table.path("expect").path("parsed").asBoolean(), rows != null, "forward " + name);
            if (rows == null) {
                continue;
            }
            List<PeerEgressSocketBinding.WindowsForwardRow> expected = new ArrayList<>();
            for (JsonNode row : table.path("expect").path("rows")) {
                expected.add(new PeerEgressSocketBinding.WindowsForwardRow(row.path("interfaceIndex").asLong(),
                        row.path("prefix").asText(), row.path("nextHop").asText(), row.path("metric").asLong()));
            }
            assertEquals(expected, rows, "forward " + name);
            forwardByName.put(name, rows);
        }
        Map<String, List<PeerEgressSocketBinding.WindowsInterfaceRow>> interfacesByName = new HashMap<>();
        for (JsonNode table : windows.path("interfaceTables")) {
            String name = table.path("name").asText();
            var rows = PeerEgressSocketBinding.parseWindowsInterfaceTable(
                    HexFormat.of().parseHex(table.path("hex").asText()));
            assertEquals(table.path("expect").path("parsed").asBoolean(), rows != null, "interfaces " + name);
            if (rows == null) {
                continue;
            }
            List<PeerEgressSocketBinding.WindowsInterfaceRow> expected = new ArrayList<>();
            for (JsonNode row : table.path("expect").path("rows")) {
                expected.add(new PeerEgressSocketBinding.WindowsInterfaceRow(row.path("interfaceIndex").asLong(),
                        row.path("metric").asLong(), row.path("connected").asBoolean(),
                        row.path("disableDefaultRoutes").asBoolean()));
            }
            assertEquals(expected, rows, "interfaces " + name);
            interfacesByName.put(name, rows);
        }

        JsonNode joined = windows.path("routes");
        var forward = forwardByName.get(joined.path("forward").asText());
        var interfaces = interfacesByName.get(joined.path("interfaces").asText());
        assertNotNull(forward);
        assertNotNull(interfaces);
        var routes = PeerEgressSocketBinding.windowsRoutes(forward, interfaces);
        assertEquals(routes(joined.path("expect")), routes, "joined routes");
        for (JsonNode testCase : windows.path("cases")) {
            assertChoice("windows " + testCase.path("name").asText(), routes, testCase);
        }
    }

    @Test
    void macosTablesMatchTheSharedVector() throws IOException {
        JsonNode macos = vector().path("macos");
        Map<String, List<PeerEgressSocketBinding.Route>> byTable = new HashMap<>();
        for (JsonNode entry : macos.path("routes")) {
            String table = entry.path("table").asText();
            var routes = PeerEgressSocketBinding.macosRoutes(macos.path("tables").path(table).asText());
            assertEquals(routes(entry.path("expect")), routes, "macos routes " + table);
            byTable.put(table, routes);
        }
        for (JsonNode testCase : macos.path("cases")) {
            assertChoice("macos " + testCase.path("name").asText(),
                    byTable.get(testCase.path("table").asText()), testCase);
        }
    }

    @Test
    void linuxTablesMatchTheSharedVector() throws IOException {
        JsonNode linux = vector().path("linux");
        assertTrue(linux.path("routes").size() > 0, "no Linux tables");
        for (JsonNode entry : linux.path("routes")) {
            String table = entry.path("table").asText();
            assertEquals(routes(entry.path("expect")),
                    PeerEgressRouteCommands.parseRouteTable(linux.path("tables").path(table).asText()),
                    "linux routes " + table);
        }
        // The captures that justify reading the table at all: `ip route get` answering with the tunnel.
        linux.path("captures").fields().forEachRemaining(capture -> {
            if (capture.getKey().startsWith("get-covered")) {
                PeerEgressRouteCommands.Hop hop =
                        PeerEgressRouteCommands.parseRouteGet(capture.getValue().path("stdout").asText());
                assertTrue(PeerEgressRouteCommands.hopIsDevice(hop, "specus0"),
                        capture.getKey() + " parsed as " + hop);
            }
        });
    }

    @Test
    void bypassHopsMatchTheSharedVector() throws IOException {
        JsonNode root = vector();
        Map<String, Map<String, List<PeerEgressSocketBinding.Route>>> tables = new HashMap<>();
        tables.put("linux", new HashMap<>());
        tables.put("macos", new HashMap<>());
        tables.put("windows", new HashMap<>());
        root.path("linux").path("tables").fields().forEachRemaining(table -> tables.get("linux")
                .put(table.getKey(), PeerEgressRouteCommands.parseRouteTable(table.getValue().asText())));
        root.path("macos").path("tables").fields().forEachRemaining(table -> tables.get("macos")
                .put(table.getKey(), PeerEgressSocketBinding.macosRoutes(table.getValue().asText())));
        JsonNode windows = root.path("windows");
        List<PeerEgressSocketBinding.WindowsForwardRow> forward = null;
        List<PeerEgressSocketBinding.WindowsInterfaceRow> interfaces = null;
        for (JsonNode table : windows.path("forwardTables")) {
            if (table.path("name").asText().equals(windows.path("routes").path("forward").asText())) {
                forward = PeerEgressSocketBinding.parseWindowsForwardTable(
                        HexFormat.of().parseHex(table.path("hex").asText()));
            }
        }
        for (JsonNode table : windows.path("interfaceTables")) {
            if (table.path("name").asText().equals(windows.path("routes").path("interfaces").asText())) {
                interfaces = PeerEgressSocketBinding.parseWindowsInterfaceTable(
                        HexFormat.of().parseHex(table.path("hex").asText()));
            }
        }
        assertNotNull(forward);
        assertNotNull(interfaces);
        tables.get("windows").put("typical", PeerEgressSocketBinding.windowsRoutes(forward, interfaces));

        JsonNode cases = root.path("hops").path("cases");
        assertTrue(cases.size() > 0, "no hop cases");
        for (JsonNode testCase : cases) {
            String name = testCase.path("name").asText();
            String platform = testCase.path("platform").asText();
            List<PeerEgressSocketBinding.Route> routes = "inline".equals(platform)
                    ? routes(testCase.path("routes"))
                    : tables.get(platform).get(testCase.path("table").asText());
            List<String> owned = new ArrayList<>();
            testCase.path("owned").forEach(prefix -> owned.add(prefix.asText()));
            PeerEgressSocketBinding.Route hop = PeerEgressSocketBinding.selectBypassHop(routes,
                    testCase.path("tunnel").asText(), owned, testCase.path("destination").asText());
            JsonNode expected = testCase.path("expect").path("hop");
            if (expected.isNull()) {
                assertNull(hop, name + ": want no hop");
            } else {
                assertNotNull(hop, name + ": no hop");
                assertEquals(expected.path("interface").asText(), hop.iface(), name + " interface");
                assertEquals(expected.path("gateway").asText(), hop.gateway(), name + " gateway");
            }
        }
    }

    /**
     * The fallback a commander takes when its query answers with the tunnel: the table with the
     * tunnel left out, and a refusal that says why when nothing else leads there.
     */
    /** Every sampled table in the vector, as this runtime parses it. */
    private static Map<String, Map<String, List<PeerEgressSocketBinding.Route>>> tables(JsonNode root) {
        Map<String, Map<String, List<PeerEgressSocketBinding.Route>>> tables = new HashMap<>();
        tables.put("linux", new HashMap<>());
        tables.put("macos", new HashMap<>());
        tables.put("windows", new HashMap<>());
        root.path("linux").path("tables").fields().forEachRemaining(table -> tables.get("linux")
                .put(table.getKey(), PeerEgressRouteCommands.parseRouteTable(table.getValue().asText())));
        root.path("macos").path("tables").fields().forEachRemaining(table -> tables.get("macos")
                .put(table.getKey(), PeerEgressSocketBinding.macosRoutes(table.getValue().asText())));
        JsonNode windows = root.path("windows");
        List<PeerEgressSocketBinding.WindowsForwardRow> forward = null;
        List<PeerEgressSocketBinding.WindowsInterfaceRow> interfaces = null;
        for (JsonNode table : windows.path("forwardTables")) {
            if (table.path("name").asText().equals(windows.path("routes").path("forward").asText())) {
                forward = PeerEgressSocketBinding.parseWindowsForwardTable(
                        HexFormat.of().parseHex(table.path("hex").asText()));
            }
        }
        for (JsonNode table : windows.path("interfaceTables")) {
            if (table.path("name").asText().equals(windows.path("routes").path("interfaces").asText())) {
                interfaces = PeerEgressSocketBinding.parseWindowsInterfaceTable(
                        HexFormat.of().parseHex(table.path("hex").asText()));
            }
        }
        assertNotNull(forward);
        assertNotNull(interfaces);
        tables.get("windows").put("typical", PeerEgressSocketBinding.windowsRoutes(forward, interfaces));
        return tables;
    }

    /**
     * What the consumer's own routes look like after a network change, and the repair each needs.
     * The policy -- a tunnel route on another interface is a conflict, a bypass follows the hop the
     * table would choose with the owned prefixes left out -- has to be the same in three runtimes.
     */
    @Test
    void driftMatchesTheSharedVector() throws IOException {
        JsonNode root = vector();
        Map<String, Map<String, List<PeerEgressSocketBinding.Route>>> tables = tables(root);
        JsonNode cases = root.path("drift").path("cases");
        assertTrue(cases.size() > 0, "no drift cases");
        for (JsonNode testCase : cases) {
            String name = testCase.path("name").asText();
            String platform = testCase.path("platform").asText();
            List<PeerEgressSocketBinding.Route> routes = "inline".equals(platform)
                    ? routes(testCase.path("routes"))
                    : tables.get(platform).get(testCase.path("table").asText());
            List<PeerEgressRoutePlanner.Route> owned = new ArrayList<>();
            for (JsonNode entry : testCase.path("owned")) {
                owned.add(new PeerEgressRoutePlanner.Route(entry.path("cidr").asText(),
                        PeerEgressRoutePlanner.Kind.fromWireName(entry.path("kind").asText()), ""));
            }
            List<PeerEgressSocketBinding.Drift> found =
                    PeerEgressSocketBinding.drifts(owned, routes, testCase.path("tunnel").asText());
            JsonNode expect = testCase.path("expect");
            assertEquals(expect.size(), found.size(), name + ": found " + found);
            for (int index = 0; index < expect.size(); index++) {
                JsonNode want = expect.get(index);
                PeerEgressSocketBinding.Drift got = found.get(index);
                assertEquals(want.path("cidr").asText(), got.route().cidr(), name + ": drift " + index);
                assertEquals(want.path("reason").asText(), got.reason(), name + ": drift " + index + " reason");
                assertEquals(want.path("action").asText(), got.action(), name + ": drift " + index + " action");
            }
        }
    }

    @Test
    void theHopFromTheTableLeavesTheTunnelOut() throws IOException {
        List<PeerEgressSocketBinding.Route> routes = PeerEgressRouteCommands.parseRouteTable(
                vector().path("linux").path("tables").path("show-rich").asText());

        PeerEgressSocketBinding.Route hop =
                PeerEgressSocketBinding.bypassHopFromTable("203.0.113.9", "specus0", () -> routes);
        assertEquals("eth0", hop.iface());
        assertEquals("192.168.64.1", hop.gateway());
        assertThrows(PeerEgressSocketBinder.NoPhysicalRouteException.class,
                () -> PeerEgressSocketBinding.bypassHopFromTable("192.0.2.9", "specus0", () -> routes));
        IOException unreadable = assertThrows(IOException.class,
                () -> PeerEgressSocketBinding.bypassHopFromTable("203.0.113.9", "specus0", () -> {
                    throw new IOException("ip: not found");
                }));
        assertTrue(unreadable.getMessage().contains("ip: not found"), unreadable.getMessage());
    }

    @Test
    void socketOptionsMatchTheSharedVector() throws IOException {
        JsonNode options = vector().path("socketOptions");
        assertEquals(PeerEgressSocketBinding.WINDOWS_IP_UNICAST_IF, options.path("windows").path("name").asInt());
        assertEquals(PeerEgressSocketBinding.MACOS_IP_BOUND_IF, options.path("macos").path("name").asInt());
        assertEquals(PeerEgressSocketBinding.IPPROTO_IP, options.path("windows").path("level").asInt());
        assertEquals(PeerEgressSocketBinding.IPPROTO_IP, options.path("macos").path("level").asInt());
        for (JsonNode testCase : options.path("windows").path("cases")) {
            assertEquals(testCase.path("hex").asText(), HexFormat.of().formatHex(
                    PeerEgressSocketBinding.windowsUnicastInterfaceOption(testCase.path("index").asLong())),
                    "IP_UNICAST_IF " + testCase.path("index").asLong());
        }
        for (JsonNode testCase : options.path("macos").path("cases")) {
            assertEquals(testCase.path("hex").asText(), HexFormat.of().formatHex(
                    PeerEgressSocketBinding.macosBoundInterfaceOption(testCase.path("index").asLong())),
                    "IP_BOUND_IF " + testCase.path("index").asLong());
        }
    }
}
